[doc] C-API documentation & generator (#5736)
taichi-dev/taichi: c_api/docs/taichi/taichi_core.h.md

---
sidebar_position: 1
---

# Core Functionality

Taichi Core exposes all necessary interfaces to offload AOT modules to Taichi. Here lists the features universally available disregards to any specific backend. These APIs are still in active development so is subject to change.

## Availability

> Taichi C-API has bridged the following backends:

Reviewer: The word "bridge" does not sound very clear to me. How about "is supported on the following backends"? Set me straight if I were wrong.

|Backend|Offload Target|Maintenance Tier|
|-|-|-|
|Vulkan|GPU|Tier 1|
|CUDA (LLVM)|GPU (NVIDIA)|Tier 1|
|CPU (LLVM)|CPU|Tier 1|
|DirectX 11|GPU (Windows)|N/A|
|Metal|GPU (macOS, iOS)|N/A|
|OpenGL|GPU|N/A|

> The backends with tier 1 support are the most intensively developed and tested ones. In contrast, you would expect a delay in fixes against minor issues on tier 2 backends. The backends currently unsupported might become supported. Among all the tier 1 backends, Vulkan has the most outstanding cross-platform compatibility, so most of the new features will be first available on Vulkan.

Reviewer: Tier-2 backends were not listed. I could only see "N/A".

> For convenience, in the following text (and other C-API documentations), the term **host** refers to the user of the C-API; the term **device** refers to the logical (conceptual) compute device that Taichi Runtime offloads its compute tasks to. A *device* might not be an actual discrete processor away from the CPU and the *host* MAY NOT be able to access the memory allocated on the *device*.

Reviewer: ```suggestion
For convenience, in the following text and other C-API documents, the term *host* refers to the user of the C-API; the term *device* refers to the logical (conceptual) compute device, to which Taichi's runtime offloads its compute tasks. A *device* may not be a real-world discrete processor other than the CPU and the *host* may *not* be able to access the memory allocated to the *device*.
```

Reviewer: I'm not quite certain about this editing.

> Unless explicitly explained, **device**, **backend**, **offload targer** and **GPU** are used interchangeably; **host**, **user code**, **user procedure** and **CPU** are used interchangeably too.

Reviewer: ```suggestion
Unless otherwise specified, **device**, **backend**, **offload targer**, and **GPU** are interchangeable; **host**, **user code**, **user procedure**, and **CPU** are interchangeable.
```

## How to...

> In this section we give an brief introduction about what you might want to do with the Taichi C-API.

Reviewer: ```suggestion
The following section provides a brief introduction to the Taichi C-API.
```

mai.size = 1024; // Size in bytes.
mai.usage = TI_MEMORY_USAGE_STORAGE_BIT;
TiMemory memory = ti_allocate_memory(runtime, &mai);
```

You MAY free allocated memory explicitly; but memory allocations will be automatically freed when the related `handle.runtime` is destroyed.

```cpp
ti_free_memory(runtime, memory);
```

### Allocate Host-Accessible Memory

By default, memory allocations are physically or conceptually local to the offload target for performance reasons. You can configure the allocate info to enable host access to memory allocations. But please note that host-accessible allocations MAY slow down computation on GPU because of the limited bus bandwidth between the host memory and the device.

To allow data to be streamed into the memory, `host_write` MUST be set true.

```cpp
TiMemoryAllocateInfo mai {};
mai.size = 1024; // Size in bytes.
mai.host_write = true;
mai.usage = TI_MEMORY_USAGE_STORAGE_BIT;
TiMemory steaming_memory = ti_allocate_memory(runtime, &mai);

// ...

std::vector<uint8_t> src = some_random_data_source();

void* dst = ti_map_memory(runtime, steaming_memory);
std::memcpy(dst, src.data(), src.size());
ti_unmap_memory(runtime, streaming_memory);
```

> To read data back to the host, `host_read` MUST be set true.

Reviewer: The word "read" does not sound about right to me... "Write"? Maybe it's just that I'm new to it.

```cpp
TiMemoryAllocateInfo mai {};
mai.size = 1024; // Size in bytes.
mai.host_write = true;
mai.usage = TI_MEMORY_USAGE_STORAGE_BIT;
TiMemory read_back_memory = ti_allocate_memory(runtime, &mai);

// ...

std::vector<uint8_t> dst(1024);
void* src = ti_map_memory(runtime, read_back_memory);
std::memcpy(dst.data(), src, dst.size());
ti_unmap_memory(runtime, read_back_memory);

ti_free_memory(runtime, read_back_memory);
```

> **NOTE** `host_read` and `host_write` can be set true simultaneously.

Reviewer: ```suggestion
:::note
You can set `host_read` and `host_write` at the same time.
:::
```

ndarray.elem_type = TI_DATA_TYPE_F32;

std::array<TiArgument, 3> args{};

TiArgument& arg0 = args[0];
arg0.type = TI_ARGUMENT_TYPE_I32;
arg0.value.i32 = 123;

TiArgument& arg1 = args[1];
arg1.type = TI_ARGUMENT_TYPE_F32;
arg1.value.f32 = 123.0f;

TiArgument& arg2 = args[2];
arg1.type = TI_ARGUMENT_TYPE_NDARRAY;
arg1.value.ndarray = ndarray;

ti_launch_kernel(runtime, kernel, args.size(), args.data());
```

You can launch a compute graph in a similar way. But additionally please ensure the argument names matches those in the Python source.

```cpp
std::array<TiNamedArgument, 3> named_args{};
TiNamedArgument& named_arg0 = named_args[0];
named_arg0.name = "foo";
named_arg0.argument = args[0];
TiNamedArgument& named_arg1 = named_args[1];
named_arg1.name = "bar";
named_arg1.argument = args[1];
TiNamedArgument& named_arg2 = named_args[2];
named_arg2.name = "baz";
named_arg2.argument = args[2];

ti_launch_compute_graph(runtime, compute_graph, named_args.size(), named_args.data());
```

When you have launched all kernels and compute graphs for this batch, you should `function.submit` and `function.wait` for the execution to finish.

```cpp
ti_submit(runtime);
ti_wait(runtime);
```

> **WARNING** This part is subject to change. We're gonna introduce multi-queue in the future.

Reviewer: - "gonna" is too colloquial. 
- "We're" is incorrect. Ought to be "We are".

> **WARNING** This part is subject to change. We're gonna introduce multi-queue in the future.

Reviewer: ```suggestion
**WARNING** This part is subject to change. We will introduce multi-queue in the future.
```

`definition.false`

A condition or a predicate is not satisfied; a statement is invalid.

`alias.flags`

A bit field that can be used to represent 32 orthogonal flags. Bits unspecified in the corresponding flag enum are ignored.

**NOTE** Enumerations and bit-field flags in the C-API have a `TI_XXX_MAX_ENUM` case to ensure the enum to have a 32-bit range and in-memory size. It has no semantical impact and can be safely ignored.

`definition.null_handle`

A sentinal invalid handle that will never be produced from a valid call to Taichi C-API.

`handle.runtime`

Taichi runtime represents an instance of a logical backend and its internal dynamic state. The user is responsible to synchronize any use of `handle.runtime`. The user MUST NOT manipulate multiple `handle.runtime`s in a same thread.

`handle.aot_module`

An ahead-of-time (AOT) compiled Taichi module, which contains a collection of kernels and compute graphs.

`handle.event`

A synchronization primitive to manage on-device execution flows in multiple queues.

`handle.memory`

A contiguous allocation of on-device memory.

`handle.kernel`

A Taichi kernel that can be launched on device for execution.

`handle.compute_graph`

> A collection of Taichi kernels (a compute graph) to be launched on device in predefined order.

Reviewer: ```suggestion
A collection of Taichi kernels (a compute graph) to launch on device in a predefined order.
```

- `structure.nd_array.elem_shape`: Shape of the ND-array elements. You usually need to set this if it's a vector or matrix ND-array.
- `structure.nd_array.elem_type`: Primitive data type of the ND-array elements.

`union.argument_value`

A scalar or structured argument value.

- `union.argument_value.i32`: Value of a 32-bit one's complement signed integer.
- `union.argument_value.f32`: Value of a 32-bit IEEE 754 floating-poing number.
- `union.argument_value.ndarray`: An ND-array to be bound.

`structure.argument`

An argument value to feed kernels.

- `structure.argument.type`: Type of the argument.
- `structure.argument.value`: Value of the argument.

`structure.named_argument`

An named argument value to feed compute graphcs.

- `structure.named_argument.name`: Name of the argument.
- `structure.named_argument.argument`: Argument body.

`function.create_runtime`

Create a Taichi Runtime with the specified `enumeration.arch`.

`function.destroy_runtime`

Destroy a Taichi Runtime.

`function.allocate_memory`

Allocate a contiguous on-device memory with provided parameters.

`function.free_memory`

Free a memory allocation.

`function.map_memory`

> Map an on-device memory to a host-addressible space. The user MUST ensure the device is not being used by any device command before the map.

Reviewer: ```suggestion
Maps an on-device memory to a host-addressible space. You *must* ensure that the device is not being used by any device command before the mapping.
```
Always address your readers with "you". 
Device or on-device memory?

`function.unmap_memory`

Unmap an on-device memory and make any host-side changes about the memory visible to the device. The user MUST ensure there is no further access to the previously mapped host-addressible space.

`function.create_event`

> Create an event primitive.

Reviewer: ```suggestion
Creates an event primitive.
```

`function.destroy_event`

> Destroy an event primitive.

Reviewer: ```suggestion
Destroys an event primitive.
```

`function.copy_memory_device_to_device`

Copy the content of a contiguous subsection of on-device memory to another. The two subsections MUST NOT overlap.

`function.launch_kernel`

Launch a Taichi kernel with provided arguments. The arguments MUST have the same count and types in the same order as in the source code.

`function.launch_compute_graph`

> Launch a Taichi compute graph with provided named arguments. The named arguments MUST have the same count, names and types as in the source code.

Reviewer: ```suggestion
Launches a Taichi compute graph with provided named arguments. The named arguments *must* have the same count, names, and types as in the source code.
```

`function.signal_event`

> Set an event primitive to a signaled state, so the queues waiting upon the event can go on execution. If the event has been signaled before, the event MUST be reset with `function.reset_event`; otherwise it is an undefined behavior.

Reviewer: ```suggestion
Sets an event primitive to a signaled state so that the queues waiting for it can go on execution. If the event has been signaled, you *must* call `function.reset_event` to reset it; otherwise, an undefined behavior would occur.
```

`function.reset_event`

> Set a signaled event primitive back to an unsignaled state.

Reviewer: ```suggestion
Sets a signaled event primitive back to an unsignaled state.
```

`function.wait_event`

> Wait on an event primitive until it transitions to a signaled state. The user MUST signal the awaited event; otherwise it is an undefined behavior.

Reviewer: ```suggestion
Waits until an event primitive transitions to a signaled state. You *must* signal the awaited event; otherwise, an undefined behavior would occur.
```
You *must* signal the awaited event. --> Signal the awaited event before calling this method??
